### ISSUE TITLE
Does not work with NixOS

### ISSUE THREAD

**Issue by WithoutCaps** [google/protobuf-gradle-plugin#426]
Since NixOS handles packages quite differently and does not depend on standard linux paths like /lib, /bin, etc - the downloaded binary does not execute. Exception:
```
Execution failed for task ':rpc:generateProto'.
> java.io.IOException: Cannot run program "/home/gytis/.local/share/gradle/caches/modules-2/files-2.1/com.google.protobuf/protoc/3.6.1/f8584d1e86d1adb61d9c559e4b49cedbc97c4d78/protoc-3.6.1-linux-x86_64.exe": error=2, No such file or directory
```

Temp workaround:
Install `protobuf` package manually and define this in your config:
```
   protoc {
        path = 'protoc'
    }
```

**Comment by voidzcy** [comment]
The published protoc executable in maven central is only for window, linux and mac, none of them works for NixOS. You do need to use a locally installed protoc.

**Comment by WithoutCaps** [comment]
NixOS is a Linux distrubution 
https://nixos.org/

**Comment by voidzcy** [comment]
Hmm.. Alright, that makes sense for the executable to be `protoc-3.6.1-linux-x86_64.exe`, which is weird. The os_detector detects the system is linux but still append the default extension `exe`.

This is related to #329.

You can append an `@` at the end of your protoc artifact. For example, if you are currently doing 
```
protobuf {
  // Configure the protoc executable
  protoc {
    // Download from repositories
    artifact = 'com.google.protobuf:protoc:3.0.0'
  }
  ...
}
```

It becomes 
```
protobuf {
  // Configure the protoc executable
  protoc {
    // Download from repositories
    artifact = 'com.google.protobuf:protoc:3.0.0@'
  }
  ...
}
```

**Comment by WithoutCaps** [comment]
`exe` part is just part of the name, for Linux it does not mean anything. Also, it detects that it is Linux - it states it in the name itself...

I tried appending `@`. It had no effect. (I also cleared caches and stuff)

The issue probably is with dynamic libraries linking.

**Comment by ejona86** [comment]
@voidzcy, why do you say the exe extension is involved? ["protoc-3.6.1-linux-x86_64.exe" totally exists](https://repo1.maven.org/maven2/com/google/protobuf/protoc/3.6.1/).

It isn't clear to me why NixOS causes "No such file or directory". @WithoutCaps, at what folder does "/home/gytis/.local/share/gradle/caches/modules-2/files-2.1/com.google.protobuf/protoc/3.6.1/f8584d1e86d1adb61d9c559e4b49cedbc97c4d78/protoc-3.6.1-linux-x86_64.exe" no longer exist?

I'm very surprised (although in a happy way) to see ~/.local/share/gradle; I can't get gradle to use any directory other than ~/.gradle. @WithoutCaps, do you have a local symlink in place?

I'd expect any error regarding "not depend on standard linux paths like /lib" to be because the dynamic linker can't be found: "/lib64/ld-linux-x86-64.so.2".

I don't know NixOS enough, but it would seem likely that it isn't possible to provide pre-built elf binaries for the platform since we wouldn't know where ld is located. Is that true? If so, installing the binary using the package manager is the only available approach.

**Comment by ejona86** [comment]
Oh! I know why we see " No such file or directory". Because /lib64/ld-linux-x86-64.so.2 can't be found. Because Linux loads /lib64/ld-linux-x86-64.so.2 as the binary itself (similar to how shell scripts use #! to load /usr/bin/sh).

@WithoutCaps, where is ld located in your system? It seems likely there would be a separate version for each glibc; is that right?

**Comment by voidzcy** [comment]
Oops, sorry. I thought the error message was for the protoc executable file not exists in local maven. My bad.

I found https://github.com/scalapb/ScalaPB/issues/505#issuecomment-549123471 might explains this problem, for which some other protoc users encountered in NixOS.

**Comment by WithoutCaps** [comment]
So I had a big long chat with some smart NixOS guys and basically this is the result:
```
There is no way to specify the interpreter in an ELF file that will work on nixOS and a system not having nix installed. What the plugins author could do though, is to either make the executable to use overridable via a config mechanism that does not rely on the project description, or they could identify nixos as nixos and instead of downloading some incompatible binary they could instantiate a derivation and run from it. That derivation could basically just pull down the incompatible binary and use autoPatchElf (or how it was called) to do the right thing.
```
Currently there are two ways to do this:
1. Using nix-shell (chroot environment). Here is a script:
```
{ pkgs ? import <nixpkgs> {} }:

(pkgs.buildFHSUserEnv {
  name = "gradle-shell";
  targetPkgs = pkgs: [ pkgs.gradle pkgs.glibc ];
  runScript = "zsh";
}).env
```
2. Pach binary manually
In my case that would be this:
```
patchelf --interpreter /nix/store/xg6ilb9g9zhi2zg1dpi4zcp288rhnvns-glibc-2.30/lib/ld-linux-x86-64.so.2 /home/gytis/.local/share/gradle/caches/modules-2/files-2.1/com.google.protobuf/protoc/3.6.1/f8584d1e86d1adb61d9c559e4b49cedbc97c4d78/protoc-3.6.1-linux-x86_64.exe
```

What else possibly you guys could do is add another option which would be like "prefer `/protobuf/protoc/path` over artifact if exists"

I am fine with closing this issue. Should I?

**Comment by voidzcy** [comment]
Yes, feel free to close the issue.

I am ok with the option of preferring the local installation over the artifact when both are provided in the `protoc` configuration block. Most users should just specify only one. There might be some concern that the artifact is more likely to be for a higher version as upgrading a local installed protoc is more involved than specifying a newer version of artifact to download. But if we document it clearly, it should be fine.

**Comment by WithoutCaps** [comment]
Yeah, I think I'll do a PR for you guys, probably sometime this week.

**Comment by sundbry** [comment]
We have the same issue on GNU Guix (It uses the same mechanism as NixOS for the linker).

I would prefer to have a java system property we can set ( -DprotocExecutable=...) as I think putting it in the project's build.gradle would not be very portable for different developers.

**Comment by gytis-ivaskevicius** [comment]
Okay, I kinda forgot about this issue, but I found a different solution:
```
  system.activationScripts.ldso = lib.stringAfter [ "usrbinenv" ] ''
    mkdir -m 0755 -p /lib64
    ln -sfn ${pkgs.glibc.out}/lib64/ld-linux-x86-64.so.2 /lib64/ld-linux-x86-64.so.2.tmp
    mv -f /lib64/ld-linux-x86-64.so.2.tmp /lib64/ld-linux-x86-64.so.2 # atomically replace
  '';
```
it creates `/lib64/ld-linux-x86-64.so.2` symlink which allows me to run most arbitrary binaries without patching anything
